calico_node/startup/startup.go

```diff
@@ -190,26 +190,33 @@ func configureLogging() {
 // determineNodeName is called to determine the node name to use for this instance
 // of calico/node.
 func determineNodeName() string {
-	// Determine the name of this node.
-	nodeName := os.Getenv("NODENAME")
-	if nodeName == "" {
-		// NODENAME not specified, check HOSTNAME (we maintain this for
-		// backwards compatibility).
-		log.Info("NODENAME environment not specified - check HOSTNAME")
-		nodeName = os.Getenv("HOSTNAME")
-	}
-	if nodeName == "" {
-		// The node name has not been specified.  We need to use the OS
-		// hostname - but should warn the user that this is not a
-		// recommended way to start the node container.
-		var err error
-		if nodeName, err = names.Hostname(); err != nil {
-			log.WithError(err).Error("Unable to determine hostname")
-			terminate()
-		}
-		log.Warn("Auto-detecting node name. It is recommended that an explicit value is supplied using the NODENAME environment variable.")
+	var nodeName string
+	var err error
+
+	// Determine the name of this node.  Precedence is NODENAME, HOSTNAME, os.Hostname.
+	// We use the names.Hostname which lowercases and trims the name.
+	if nodeName = strings.TrimSpace(os.Getenv("NODENAME")); nodeName != "" {
+		log.Infof("Using NODENAME environment for node name")
+	} else if nodeName = strings.TrimSpace(os.Getenv("HOSTNAME")); nodeName != "" {
+		log.Infof("Using HOSTNAME environment for node name")
+	} else if nodeName, err = names.Hostname(); err != nil {
+		log.WithError(err).Error("Unable to determine hostname")
+		terminate()
+	} else {
+		log.Warn("Using auto-detected node name. It is recommended that an explicit value is supplied using " +
+			"the NODENAME environment variable.")
 	}
-	return nodeName
+
+	// Ensure the node name is lowercase and has no leading/trailing whitespace.  We'll log
+	// if the name is actually modified by the case conversion.
+	finalNodeName := strings.ToLower(nodeName)
```

Comment:
I'm conflicted on this one... 

On the one hand I'm a bit loathe to change something that was provided by the user here... if they've given us a hostname with capital letters then who are we to convert that to lowercase?

On the other hand, hostnames are case-insensitive, so it should be OK. However... this isn't a "hostname" - it's a Calico nodename, which is something subtly different from a hostname even though it has the same validation requirements. 

hmmm

Reply:
I guess it's Ok, and we'll see if anyone complains about it.

Reply:
OK. I think we agreed that:

- NODENAME should not be altered
- HOSTNAME should be automatically lowercased 
- names.Hostname() already handles casing.